### PR TITLE
Resize cubic and lanczos interpolation

### PR DESCRIPTION
**Pull request by fiona-gladwin**
- Add cubic and lanczos interpolation support for both the HOST and HIP backend.
- Add support for F32, F16, U8 and I8 data types of planar and packed versions.
- Modify rpp-unittests and rpp-performancetests accordingly.

### CODE REVIEW

**Review by r-abishek — CHANGES_REQUESTED** [review]
I have similar comments on the GPU side too, but haven't looked in detail. Will take another look after a round of changes.

File: src/include/cpu/rpp_cpu_common.hpp [hunk 1]
```diff
@@ -82,6 +83,53 @@ inline int power_function(int a, int b)
...
+}
+
+inline float sinc(float x) {
+    x *= M_PI;
+    if (std::abs(x) < 1e-5f)
+    return 1.0f - x * x * (1.0f / 6);  // remove singularity by using Taylor expansion
+    return std::sin(x) / x;
+}
+
+inline void CalculateLanczosCoefficients(float* coeffs, float x, int a)
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_common.hpp:109]: This one needs to be optimized too. If 'a' is a pre-defined constant as defined in L1137, and you only need [a, 2 * a, 1 / a, a - 1] inside this functionality, please compute them in an array say aParams[4] or lanczosParams[4] and reuse them here. Currently, these 4 constant values are being computed width*height times for every single function call

File: src/include/cpu/rpp_cpu_common.hpp [hunk 2]
```diff
@@ -21,13 +21,14 @@ typedef halfhpp Rpp16f;
 #define RPPMAX2(a,b)                    ((a > b) ? a : b)
 #define RPPMAX3(a,b,c)                  ((a > b) && (a > c) ?  a : ((b > c) ? b : c))
 #define RPPINRANGE(a, x, y)             ((a >= x) && (a <= y) ? 1 : 0)
-#define RPPPRANGECHECK(value, a, b)     (value < (Rpp32f) a) ? ((Rpp32f) a) : ((value < (Rpp32f) b) ? value : ((Rpp32f) b))
+#define RPPPRANGECHECK(value, a, b)     ((value < a) ? a : ((value < b) ? value : b))
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply src/include/cpu/rpp_cpu_common.hpp:24]: Yes erase and water unit tests pass with this mod

File: src/include/cpu/rpp_cpu_common.hpp [hunk 3]
```diff
@@ -82,6 +83,53 @@ inline int power_function(int a, int b)
     return product;
 }
 
+inline void saturate_pixel(Rpp32f pixel, Rpp8u* dst) {
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply src/include/cpu/rpp_cpu_common.hpp]: Done

File: src/include/cpu/rpp_cpu_common.hpp [hunk 4]
```diff
@@ -897,50 +916,229 @@ inline RppStatus resize_kernel_host(T* srcPtr, RppiSize srcSize, U* dstPtr, Rppi
...
+                        CalculateCubicCoefficients(coeffs_y, weightedHeight);
+                        for (int j = 0; j < dstSize.width; j++)
+                        {
+                            srcLocationColumn = ((Rpp32f) j + 0.5f) * wRatio - 0.5f;
+                            srcLocationColumnFloor = (Rpp32s) RPPFLOOR(srcLocationColumn);
+                            Rpp32f weightedWidth = srcLocationColumn - srcLocationColumnFloor;
+                            srcLocationColumnFloor = (srcLocationColumnFloor > widthLimit) ? widthLimit : srcLocationColumnFloor;
+                            CalculateCubicCoefficients(coeffs_x, weightedWidth);
+                            pixels[0] = pixels[1] = pixels[2] = pixels[3] = 0;
+                            for(int k=0; k < kernelSize; k++)
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply src/include/cpu/rpp_cpu_common.hpp]: Done

File: src/include/cpu/rpp_cpu_common.hpp [hunk 5]
```diff
@@ -897,50 +916,229 @@ inline RppStatus resize_kernel_host(T* srcPtr, RppiSize srcSize, U* dstPtr, Rppi
...
+        Rpp32s kernelSize = 4;
+        Rpp32s kernelSize2 = kernelSize/2;
+        T *srcPtrTemp, *srcPtrRow0, *srcPtrRow1, *srcPtrRow2, *srcPtrRow3;
+        Rpp32f coeffs_x[kernelSize], coeffs_y[kernelSize], pixels[kernelSize];
+        U *dstPtrTemp;
+        srcPtrTemp = srcPtr;
+        dstPtrTemp = dstPtr;
+        if (chnFormat == RPPI_CHN_PLANAR)
+        {
+            if ((typeid(Rpp16f) == typeid(T)) || (typeid(Rpp16f) == typeid(U)))
```
[the remark this answers is not in the document]
Comment (fiona-gladwin) [reply src/include/cpu/rpp_cpu_common.hpp:968]: Yes, removing the if else condition

File: src/include/cpu/rpp_cpu_common.hpp [hunk 6]
```diff
@@ -82,6 +83,59 @@ inline int power_function(int a, int b)
...
+        sum += coeffs[i];
+    }
+    sum = 1.f/sum;
+    for(int i = 0; i < k; i++ )
+        coeffs[i] *= sum;
+}
+
+inline void calculate_cubic_coefficients(Rpp32f* coeffs, Rpp32f x)
+{
+    Rpp32f x2 = x * x;
```
Comment (r-abishek) [inline src/include/cpu/rpp_cpu_common.hpp:131]: Rpp32f xo2 = 0.5f * x;
    Rpp32f xt3o2 = xo2 * 3.0f;
    Rpp32f xp2 = x * x;
    coeffs[0] = (-xo2 + 1) * xp2 - xo2;
    coeffs[1] = (xt3o2 - 2.5) * xp2 + 1;
    coeffs[2] = -(xt3o2 + 2) * xp2 + xo2;
    coeffs[3] = 1.0f - coeffs[0] - coeffs[1] - coeffs[2];

Comment (r-abishek) [reply src/include/cpu/rpp_cpu_common.hpp:131]: Please replace with this and check if unit test is passing

**Review by r-abishek — CHANGES_REQUESTED** [review]

**Pull request closed without merging**